File: prob113.py
def non_bouncy(digits):
    return increasing(digits) + decreasing(digits)

def increasing(digits):
    return map(increasing_starting_with, range(0, 10))

def increasing_starting_with(digits, n):
    if(digits == 1):
        return 1

    acc = 0
    for i in range(0, n):
        acc += (10-i)*increasing_starting_with(i)
    return acc

print increasing(2)

File: prob113.hs
import Data.Function.Memoize (memoize2)

f :: Integer -> Integer -> Integer
f 0 _ = 1
f n k = sum $ map (memo_f (n-1)) [k..9]

g :: Integer -> Integer -> Integer
g 0 _ = 1
g n k = sum $ map (memo_g (n-1)) [0..k]

memo_f = memoize2 f
memo_g = memoize2 g

increasing :: Integer -> Integer
increasing n = f n 1

decreasing :: Integer -> Integer
decreasing n = sum $ map (g (n-1)) [1..9]

non_bouncy :: Integer -> Integer
non_bouncy n = increasing n + decreasing n - 9

main :: IO ()
main = print $ sum $ map non_bouncy [1..100]
